pkg/services/blockfetcher/blockfetcher.go

```diff
@@ -489,21 +503,59 @@ func (bfs *Service) IsActive() bool {
 	return bfs.isActive.Load()
 }
 
+// retry function with exponential backoff.
+func (bfs *Service) retry(action func() error) error {
+	var (
+		err     error
+		backoff = initialBackoff
+		timer   = time.NewTimer(0)
+	)
+	defer timer.Stop()
+	for range maxRetries {
+		if err = action(); err == nil {
+			return nil
+		}
+		if !timer.Stop() {
+			<-timer.C
+		}
```

Comment:
Could you clarify why have yoy placed this `if !timer.Stop() {` block here? To me, the `select {` below should be sufficient for proper `retry` work.

Reply:
`Reset should be invoked only on stopped or expired timers with drained channels.` but you are right as its just created it is drained.

 `If a program has already received a value from t. C, the timer is known to have expired and the channel drained, so t. Reset can be used directly.` from docs: only because of `received a value from t. C` put it here to check that time.NewTimer(0) is empty.